client/state/countries/reducer.ts

```diff
@@ -0,0 +1,59 @@
+import { withStorageKey } from '@automattic/state-utils';
+import {
+	COUNTRIES_DOMAINS_UPDATED,
+	COUNTRIES_PAYMENTS_UPDATED,
+	COUNTRIES_SMS_UPDATED,
+	COUNTRIES_WOOCOMMERCE_UPDATED,
+} from 'calypso/state/action-types';
+import { combineReducers } from 'calypso/state/utils';
+import type { CountriesState } from './types';
+import type { Reducer } from 'react';
+import type { AnyAction } from 'redux';
+
+function createListReducer<
+	TType extends Exclude< keyof CountriesState, 'woocommerce' >,
+	TActionType extends string
+>( updatedActionType: TActionType ): Reducer< CountriesState[ TType ] | undefined, AnyAction > {
+	return ( state: undefined | CountriesState[ TType ] = [], action: AnyAction ) => {
+		switch ( action.type ) {
+			case updatedActionType:
+				return action.countries;
+			default:
+				return state;
+		}
+	};
+}
+
+function createObjectReducer<
+	TType extends Extract< keyof CountriesState, 'woocommerce' >,
+	TActionType extends string
+>( updatedActionType: TActionType ): Reducer< CountriesState[ TType ] | undefined, AnyAction > {
+	return ( state: undefined | CountriesState[ TType ] = {}, action: AnyAction ) => {
+		switch ( action.type ) {
+			case updatedActionType:
+				return action.countries;
+			default:
+				return state;
+		}
+	};
+}
+
+const combinedReducer = combineReducers( {
+	// Note: it would be nice to not need to specify the second generic type
+	// here, since it can be inferred from the argument, but if you pass any
+	// generics, you must pass all generics. May possibly be solved by
+	// https://github.com/microsoft/TypeScript/pull/26349
```

Comment:
Yep, I was going to comment on this. Makes the types a bit more complex than we'd like. I wonder what other alternatives we have for simplifaction through inference here.

Reply:
Why is `TActionType` needed in these functions? `updatedActionType` could just be typed string and it should all work.

Reply:
I think perhaps `TActionType` is meant to be used beyond the argument, but that's the only place it's being used right now.

Reply:
In a previous version of the reducer factory, I was using `TActionType` to type the actions, but that code has all since been removed (the actions are now untyped strings). So actually we can just remove the second generic here entirely.